File: office_server/src/main/kotlin/foatto/office/cClient.java
//package foatto.office;
//
//import foatto.core_server.app.server.cStandart;
//import foatto.core_server.app.server.column.iColumn;
//import foatto.core_server.app.server.data.DataDate;
//import foatto.core_server.app.server.data.iData;
//import foatto.core_client.app.table.CoreTableCellInfo;
//
//import java.util.GregorianCalendar;
//import java.util.HashMap;
//import java.util.HashSet;
//
//public class cClient extends cStandart {
//
//    private GregorianCalendar toDay = new GregorianCalendar();
//
//    protected StringBuilder addSQLWhere( HashSet<String> hsTableRenameList ) {
//        StringBuilder sb = super.addSQLWhere( hsTableRenameList );
//
//        mPeople m = (mPeople) model;
//        sb.append( " AND " ).append( renameTableName( hsTableRenameList, m.getColumnPeopleWorkState().getTableName() ) ).append( '.' )
//          .append( m.getColumnPeopleWorkState().getFieldName( 0 ) ).append( " = " ).append(
//            aliasConfig.getAlias().equals( "office_client_not_need" ) ? mPeople.WORK_STATE_NOT_NEED :
//            aliasConfig.getAlias().equals( "office_client_in_work" )  ? mPeople.WORK_STATE_IN_WORK :
//                                                                        mPeople.WORK_STATE_OUT_WORK );
//        if( aliasConfig.getAlias().equals( "office_client_not_need" ) )
//            //--- здесь используется именно жёсткое использование имени поля user_id,
//            //--- т.к. в режиме "клиент" вместо user_id будет использоваться manager_id
//            sb.append( " AND OFFICE_people.user_id = 0 " );
//
//        return sb;
//    }
//
//    protected CoreTableCellInfo setTableGroupColumnStyle( HashMap<iColumn,iData> hmColumnData, iColumn column, CoreTableCellInfo tci ) {
//        super.setTableGroupColumnStyle( hmColumnData, column, tci );
//
//        if( aliasConfig.getAlias().equals( "office_client_in_work" ) ) {
//            mPeople mp = (mPeople) model;
//            if( column.equals( mp.getColumnClientPlanDate() ) ) {
//                GregorianCalendar gc = ( (DataDate) hmColumnData.get( mp.getColumnClientPlanDate() ) ).getValue();
//                gc.add( GregorianCalendar.DAY_OF_MONTH, 1 );
//                if( gc.before( toDay ) ) {
//                    tci.foreColorType = CoreTableCellInfo.FORE_COLOR_TYPE_DEFINED;
//                    tci.foreColor = TABLE_CELL_FORE_COLOR_CRITICAL;
//                }
//            }
//        }
//        return tci;
//    }
//}

File: core_server/src/main/kotlin/foatto/core_server/app/system/cClient.kt
package foatto.core_server.app.system

import foatto.core.app.UP_TIME_OFFSET
import foatto.core.link.TableCell
import foatto.core.link.TableCellForeColorType
import foatto.core_server.app.iApplication
import foatto.core_server.app.server.OrgType
import foatto.core_server.app.server.cStandart
import foatto.core_server.app.server.column.iColumn
import foatto.core_server.app.server.data.DataBoolean
import foatto.core_server.app.server.data.DataDate3Int
import foatto.core_server.app.server.data.DataString
import foatto.core_server.app.server.data.iData
import java.util.*

open class cClient : cStandart() {

    override fun addSQLWhere(hsTableRenameList: Set<String>): String {
        val tn = renameTableName(hsTableRenameList, model.modelTableName)
        val mc = model as mClient

        return super.addSQLWhere(hsTableRenameList) +
            """    
                AND $tn.${mc.columnId.getFieldName()} > 0 
                AND $tn.${mc.columnRecordType.getFieldName()} = ${OrgType.ORG_TYPE_WORKER} 
                AND $tn.${mc.columnParent.getFieldName()} = ${mc.getClientParentId(application, aliasConfig.name)} 
            """
    }

    override fun getTableColumnStyle(isNewRow: Boolean, hmColumnData: Map<iColumn, iData>, column: iColumn, tci: TableCell) {
        super.getTableColumnStyle(isNewRow, hmColumnData, column, tci)

        val mc = model as mClient
        if (column == mc.columnRecordFullName) {
            val isDisabled = (hmColumnData[mc.columnDisabled] as DataBoolean).value
            val lastLogonTime = (hmColumnData[mc.columnUserLastLoginAttemptDate] as DataDate3Int).localDate.atStartOfDay(zoneId).toEpochSecond().toInt()

            //--- раскраска фона имени пользователя в зависимости от времени последнего входа в систему
            tci.foreColorType = TableCellForeColorType.DEFINED
            tci.foreColor = application.getUserNameColor(isDisabled, lastLogonTime)
        }
    }

    override fun preSave(id: Int, hmColumnData: Map<iColumn, iData>) {
        application.checkAndSetNewPassword(conn, id, hmColumnData[(model as mClient).columnUserPassword] as? DataString)
        super.preSave(id, hmColumnData)
    }

    override fun postAdd(id: Int, hmColumnData: Map<iColumn, iData>, hmOut: MutableMap<String, Any>): String? {
        val postURL = super.postAdd(id, hmColumnData, hmOut)

        val mc = model as mClient

        //--- создать запись индивидуального сдвига часового пояса
        application.saveUserProperty(
            conn = conn,
            userId = id,
            userConfig = null,
            upName = UP_TIME_OFFSET,
            upValue = (TimeZone.getDefault().rawOffset / 1000).toString()
        )

        //--- автосоздание привязки пользователя/клиента и его типовых ролей
        mc.getClientRoleIds(application, aliasConfig.name).forEach { clientRoleId ->
            val nextId = conn.getNextIntId("SYSTEM_user_role", "id")
            conn.executeUpdate(
                """
                    INSERT INTO SYSTEM_user_role( id , role_id , user_id ) 
                    VALUES ( $nextId , $clientRoleId , $id )
                """
            )
        }

        //--- обновим конфигурацию текущего пользователя (более всего необходимо обновление списка пользователей id=name)
        hmOut[iApplication.USER_CONFIG] = application.getUserConfig(conn, userConfig.userId)

        return postURL
    }

    override fun postEdit(action: String, id: Int, hmColumnData: Map<iColumn, iData>, hmOut: MutableMap<String, Any>): String? {
        val postURL = super.postEdit(action, id, hmColumnData, hmOut)

        //--- обновим конфигурацию текущего пользователя (более всего необходимо обновление списка пользователей id=name)
        hmOut[iApplication.USER_CONFIG] = application.getUserConfig(conn, userConfig.userId)

        return postURL
    }
}